File: labs-opensource/M01-llm-mental-model/solution/chat_cli.py
"""
M01 Lab - Step 5 (Stretch): CLI Chat with History — SOLUTION
=============================================================
Run: python chat_cli.py   (type 'quit' to exit)
"""

from openai import OpenAI

client = OpenAI(base_url="http://localhost:11434/v1", api_key="ollama")

conversation = []

print("Chat with Mistral! (type 'quit' to exit)\n")

while True:
    user_input = input("You: ").strip()
    if user_input.lower() in ("quit", "exit"):
        break
    if not user_input:
        continue

    conversation.append({"role": "user", "content": user_input})

    try:
        response = client.chat.completions.create(
            model="mistral",
            messages=[
                {"role": "system", "content": "You are a friendly, helpful assistant."}
            ] + conversation,
        )
        assistant_msg = response.choices[0].message.content
        conversation.append({"role": "assistant", "content": assistant_msg})
        print(f"\nMistral: {assistant_msg}\n")
    except Exception as e:
        print(f"\nError: {e}\n")
        # Remove the failed user message so conversation stays consistent
        conversation.pop()

File: labs-opensource/M01-llm-mental-model/solution/chat_cli.js
/**
 * M01 Lab - Step 5 (Stretch): CLI Chat with History — SOLUTION
 * =============================================================
 * Run: node chat_cli.js   (type 'quit' to exit)
 */

import OpenAI from "openai";
import * as readline from "node:readline/promises";

const client = new OpenAI({ baseURL: "http://localhost:11434/v1", apiKey: "ollama" });
const rl = readline.createInterface({ input: process.stdin, output: process.stdout });

const conversation = [];

console.log("Chat with Mistral! (type 'quit' to exit)\n");

while (true) {
  // Stop if stdin has ended. Redirecting input (or Ctrl-D) closes readline, and
  // asking again then throws ERR_USE_AFTER_CLOSE instead of exiting cleanly --
  // you only see it when the chat is not driven by a person typing.
  if (rl.closed) break;

  let userInput;
  try {
    userInput = (await rl.question("You: ")).trim();
  } catch {
    break;                      // stream closed mid-prompt
  }

  if (["quit", "exit"].includes(userInput.toLowerCase())) {
    rl.close();
    break;
  }
  if (!userInput) continue;

  conversation.push({ role: "user", content: userInput });

  try {
    const response = await client.chat.completions.create({
      model: "mistral",
      messages: [
        { role: "system", content: "You are a friendly, helpful assistant." },
        ...conversation,
      ],
    });
    const assistantMsg = response.choices[0].message.content;
    conversation.push({ role: "assistant", content: assistantMsg });
    console.log(`\nMistral: ${assistantMsg}\n`);
  } catch (error) {
    console.error(`\nError: ${error.message}\n`);
    // Remove the failed user message so conversation stays consistent
    conversation.pop();
  }
}
